src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Components/ComponentNodeWriter.cs

```diff
@@ -530,6 +541,50 @@ protected static void WriteGloballyQualifiedTypeName(CodeRenderingContext contex
         }
     }
 
+    protected static void WriteComponentTypeName(CodeRenderingContext context, ComponentIntermediateNode node, ReadOnlyMemory<char> nonGenericTypeName)
+    {
+        // The type name we are given may or may not be globally qualified, and we want to map it to the component start
+        // tag, which may or may not be fully qualified. ie "global::My.Fun.Component" could map to just "Component"
+
+        // If we don't have mapping info, we don't do anything interesting
+        if (node.StartTagSpan is not { } startTagSpan ||
+            startTagSpan == SourceSpan.Undefined)
```

Comment:
FWIW, since `ComponentIntermediateNode.StartTagSpan` is assigned by calling `SyntaxNodeExtensions.GetSourceSpan(SyntaxToken, RazorSourceDocument)`, I don't think this condition is ever possible.